docs: rails debug guide - add detail re: verbose logs [ci-skip] (#54066)
rails/rails: guides/source/debugging_rails_applications.md

### What is the Logger?

Rails makes use of the `ActiveSupport::Logger` class to write log information. Other loggers, such as `Log4r`, may be substituted:

```ruby

> # config/application.rb

Reviewer: Most of the time, a different logger and level is used per environment.
The default production environment config specifically [mentions](https://github.com/rails/rails/blob/86312f5dc05b96d9c1c71ef03d257c155622f00e/railties/lib/rails/generators/rails/app/templates/config/environments/production.rb.tt#L42) it.
Maybe use that instead?
```suggestion
# config/environments/production.rb
```

Author: applied

config.logger = Logger.new(STDOUT)
config.logger = Log4r::Logger.new("Application Log")
```

> Or in the `Initializer` section, add _any_ of the following:

Reviewer: I'd probably remove this line and the following example.
I don't think we should be advising configuring logger this way as the environment config seems the most common location.

Author: removed as suggested.

### What is the Logger?

Rails makes use of the `ActiveSupport::Logger` class to write log information. Other loggers, such as `Log4r`, may be substituted:

```ruby
# config/application.rb
config.logger = Logger.new(STDOUT)
config.logger = Log4r::Logger.new("Application Log")
```

Or in the `Initializer` section, add _any_ of the following:

> ```rb

Reviewer 2: This should be `ruby` for the code-fence language

Author: - I have updated this code fence and all others in the docs.
- In order to assist in mapping proposed reviews to commits, I have added this as an additional commit: https://github.com/rails/rails/pull/54066/commits/a7c3f1f522ba47d81f1b5fc6296b3ae60d8c9e17 

I propose to squash after all the reviews are completed. If this does not work for you pls LMK i will squash immediately.

Reviewer 2: Yeah, I guess it works either way but this is the only guide which uses it:

```
git grep "\`\`\`rb"
guides/source/debugging_rails_applications.md:247:```rb
guides/source/debugging_rails_applications.md:263:```rb
guides/source/debugging_rails_applications.md:352:```rb
guides/source/debugging_rails_applications.md:367:```rb
guides/source/debugging_rails_applications.md:392:```rb
guides/source/debugging_rails_applications.md:402:```rb
guides/source/debugging_rails_applications.md:425:```rb
guides/source/debugging_rails_applications.md:445:```rb
guides/source/debugging_rails_applications.md:485:```rb
guides/source/debugging_rails_applications.md:534:```rb
guides/source/debugging_rails_applications.md:553:```rb
guides/source/debugging_rails_applications.md:575:```rb
guides/source/debugging_rails_applications.md:595:```rb
guides/source/debugging_rails_applications.md:619:```rb
guides/source/debugging_rails_applications.md:638:```rb
guides/source/debugging_rails_applications.md:662:```rb
guides/source/debugging_rails_applications.md:681:```rb
guides/source/debugging_rails_applications.md:717:```rb
guides/source/debugging_rails_applications.md:751:```rb
guides/source/debugging_rails_applications.md:760:```rb
guides/source/debugging_rails_applications.md:781:```rb
```

Would you mind squashing please?

Author: squashed and forced pushed.

### What is the Logger?

Rails makes use of the `ActiveSupport::Logger` class to write log information. Other loggers, such as `Log4r`, may be substituted:

```ruby

> # config/application.rb
> # or any environment file:

Reviewer 2: ```suggestion
# config/application.rb
```

Author: Applied.

config.logger = Logger.new(STDOUT)
config.logger = Log4r::Logger.new("Application Log")
```

Or:

```rb
# any initializer file: e.g. config/initializers/log4r.rb
Rails.logger = Logger.new(STDOUT)
Rails.logger = Log4r::Logger.new("Application Log")

> ```

Reviewer 2: Do we even need this?

I think setting config in application / environment / initializers is covered elsewhere.

Author: > Do we even need this?
> I think setting config in application / environment / initializers is covered elsewhere.

Here's my reasoning:

- I made this change actually for my own benefit:
- "What is the initializer section?" I would have rather had it laid out: so I wouldn't have to even think about the answer.
- The second reason is that as a general principle, I prefer to see the the location of the markup within the code block itself, rather than in prose.

I have reverted to the old style:
```diff
- Or in the `Initializer` section, add _any_ of the following
+ Or in the `Initializer` section, add _any_ of the following:
```

The available log levels are: `:debug`, `:info`, `:warn`, `:error`, `:fatal`,
and `:unknown`, corresponding to the log level numbers from 0 up to 5,
respectively. To change the default log level:

```ruby

> # config/application.rb

Reviewer: Similarly, I think this should mention the environment file as that [is where](https://github.com/rails/rails/blob/86312f5dc05b96d9c1c71ef03d257c155622f00e/railties/lib/rails/generators/rails/app/templates/config/environments/production.rb.tt#L45) this is located in a new Rails app.
```suggestion
# config/environments/production.rb
```

Author: applied.

Author: applied.

The available log levels are: `:debug`, `:info`, `:warn`, `:error`, `:fatal`,
and `:unknown`, corresponding to the log level numbers from 0 up to 5,
respectively. To change the default log level:

```ruby

> # config/application.rb
> # or any environment file:

Reviewer 2: ```suggestion
# config/application.rb
```

Author: applied.

config.log_level = :warn

> ```
> 
> Or:
> 
> ```ruby
> # any initializer file e.g. config/initializers/logger.rb
> Rails.logger.level = 0

Reviewer 2: ```suggestion
```

=> #<Comment id: 2, author: "1", body: "Well, actually...", article_id: 1, created_at: "2018-10-19 00:56:10", updated_at: "2018-10-19 00:56:10">
```

> But it becomes manifest after enabling `verbose_query_logs`:

Reviewer 2: ```suggestion
After enabling `verbose_query_logs` we can see additional info for each query:
```

=> #<Comment id: 2, author: "1", body: "Well, actually...", article_id: 1, created_at: "2018-10-19 00:56:10", updated_at: "2018-10-19 00:56:10">
```

Below each database statement you can see arrows pointing to the specific source filename (and line number) of the method that resulted in a database call e.g. `↳ app/models/article.rb:5`.

This can help you identify and address performance problems caused by N+1 queries: i.e. single database queries that generate multiple additional queries.

> Verbose query logs are [enabled by default](/configuring.html#config-active-record-verbose-query-logs) in the development environment logs.

Reviewer 2: ```suggestion
Verbose query logs are [enabled by default](configuring.html#config-active-record-verbose-query-logs) in the development environment logs.
```

I think the slash could potentially lose any version tag that proceeds the file path (all guides are at the same directory-level).

Author: Change applied and forced-pushed: there should not be a single squashed commit.

WARNING: We recommend against using this setting in production environments. It relies on Ruby's `Kernel#caller` method which tends to allocate a lot of memory in order to generate stacktraces of method calls. Use query log tags (see below) instead.

### Verbose Enqueue Logs

Similar to the "Verbose Query Logs" above, allows to print source locations of methods that enqueue background jobs.

> Verbose enqueue logs are enabled by default in the development environment logs after Rails 7.1.

Reviewer: I find `after` a bit unclear here.
Maybe "starting with" (which we often use in the configuration guide?):
```suggestion
Verbose enqueue logs are enabled by default in the development environment logs starting with Rails 7.2.
```

Author: > So I think we can remove the reference to the version.

Removing it goes against my specific need. I had a pre-7.1 rails app. I cannot be the only one? I had gone so far as to check against the original PR. In light of the aforementioned: pls confirm whether you want this removed?

Reviewer: If we show the version it was released to every feature the guides would become unreadable.

Maybe expand the 7.1 release notes for this features with the setting name? That would make it more searchable (although search is not yet supported):
https://github.com/rails/rails/blob/e25d738430bdc6bdd04cd28be705484ea953e74e/guides/source/7_1_release_notes.md?plain=1#L869

Author: > If we show the version it was released, for every feature, the guides would become unreadable.

agreed. I have added a link instead, given it is not searchable.

> Maybe expand the 7.1 release notes for this features with the setting name? That would make it more searchable (although search is not yet supported):

I have done as suggested in principle, but altered the wording and force pushed:

`Add support for logging background job enqueue callers via `config.active_job.verbose_enqueue_logs`.`

If this is not conducive to the goals of the docs pls have no hesitation in alerting or correcting me.

```ruby

> # config/application.rb

Reviewer: If we recommend not enabling it in production, we should probably use an example that only sets it for development?
```suggestion
# config/environments/development.rb
```

Verbose query logs are enabled by default in the development environment logs after Rails 5.2.

```rb

> # ./config/environments/development.rb

Reviewer 2: I think this is redundant to the statement above:

> After running `ActiveRecord.verbose_query_logs = true` in the `bin/rails console` session to enable verbose query logs and running the method again, it becomes obvious what single line of code is generating all these discrete database calls:

Although they are changing the setting in different places, I don't see the point of having both. I would prefer to mention setting the flag in the config file rather than in a console.

Author: > Although they are changing the setting in different places, I don't see the point of having both. I would prefer to mention setting the flag in the config file rather than in a console.

Agreed: the console version is stated above quite clearly. But I didn't want to make the change in the console, I wanted to change the `development.rb` file. I had to go search out the config setting specifically, which was a minor annoyance, and the genesis for this PR.

The point you have made above seems uniform with the rest of the docs: 

![uniformity](https://github.com/user-attachments/assets/466a6130-6d4c-43bc-a504-bc06141db28a)

However, my own 2 cents: to leave the console settings, and _add_ the environmental settings. 

Please confirm whether this is the change you want?

```diff
- After running ActiveRecord.verbose_query_logs = true in the bin/rails console session to enable 
- verbose query logs and running the method again, it becomes obvious what single line of code is 
- generating all these discrete database calls:
+ # ./config/environments/development.rb
+ config.active_record.verbose_query_logs = true
```

Reviewer 2: All of the settings you highlighted are enabled by default in development for newly-generated Rails applications.

Maybe your app predates that, or we (sometimes do) change the defaults in that file.

I think something like "this is enabled by default for new Rails apps in development, but you can enable it by adding this to your config:" with a code block for the flag.

We have different wording there in each section that could all be cleaned up. I will let you decide how to phrase it.

I don't think we need to show the explicit config environment file path, but it might be useful to have both "use `ActiveRecord.whatever_class_variable = true` or via `config.active_record.whatever_thing = true`.

Sorry to sign you up for extra work, but while you're in that file would you mind tidying things up a bit? There might be other cases like that and I'd rather we make them consistent in a single pass. If you're interested, of course.

Author: > Sorry to sign you up for extra work, but while you're in that file would you mind tidying things up a bit? There might be other cases like that and I'd rather we make them consistent in a single pass. If you're interested, of 

It is my pleasure: ye have asked and ye shall receive.

I have liberally updated (hopefully) accordance with your suggestions:

1. Made concise where possible.
2. Applied file locations within markup as opposed to within prose where possible due to the above.
3. I did not see much benefit in toggling `ActiveJob.verbose_enqueue_logs = true` within the irb, but i have provided it nonetheless.
4. In terms of doing the same for the [`query_log_tags `](https://api.rubyonrails.org/classes/ActiveRecord/QueryLogs.html). I have skipped it entirely.
5. `in Initializer section` i have assumed this to mean `config/initializers`, perhaps it may refer to environment initializers.

Any issues pls LMK.

Reviewer 2: Thanks, I will try to review this soon.

In the mean time the linter is failing due to trailing spaces:

```
/rails/guides/source/debugging_rails_applications.md:113: [MD009](https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md009---trailing-spaces) Trailing spaces
/rails/guides/source/debugging_rails_applications.md:149: [MD009](https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md009---trailing-spaces) Trailing spaces
/rails/guides/source/debugging_rails_applications.md:249: [MD009](https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md009---trailing-spaces) Trailing spaces
/rails/guides/source/debugging_rails_applications.md:268: [MD009](https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md009---trailing-spaces) Trailing spaces
/rails/guides/source/debugging_rails_applications.md:289: [MD009](https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md#md009---trailing-spaces) Trailing spaces
```

Author: > In the mean time the linter is failing due to trailing spaces:

- [x] Fix: Trailing white spaces and the linter is now passing.

Pls LMK general policy re: the aforementioned changes, and I can emend accordingly if required..

Reviewer 2: Please [squash](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#squashing-commits) when you're ready for another review :bow:

Author: > Please [squash](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#squashing-commits) when you're ready for another review 🙇

It is squashed and ready for another review.

=> #<Comment id: 2, author: "1", body: "Well, actually...", article_id: 1, created_at: "2018-10-19 00:56:10", updated_at: "2018-10-19 00:56:10">
```

> Below each database statement you can see arrows pointing to the specific source filename (and line number) of the method that resulted in a database call e.g. `↳ app/models/article.rb:5`. This can help you identify and address performance problems caused by N+1 queries: single database queries that generates multiple additional queries.

Reviewer 2: ```suggestion
Below each database statement you can see arrows pointing to the specific source filename (and line number) of the method that resulted in a database call e.g. `↳ app/models/article.rb:5`.
This can help you identify and address performance problems caused by N+1 queries: single database queries that generates multiple additional queries.
```

Author: ```diff
- This can help you identify and address performance problems caused by N+1 queries: single database queries that generates multiple additional queries.
+ This can help you identify and address performance problems caused by N+1 queries: i.e. single database queries that generate multiple additional queries.
```
(small change applied: grammar + readability.)

Verbose query logs are enabled by default in the development environment logs after Rails 5.2.

```rb
# config/application.rb
# or any environment file:
config.active_record.verbose_query_logs = true
```

```irb
# bin/rails console
ActiveRecord.verbose_query_logs = true
```

> 

Reviewer 2: I think what we want to do here is just link to the config page:
`configuring.html#config-active-record-verbose-query-logs`

And we should remove the sentence above this about "after Rails 5.2" it is not relevant anymore.

The important thing is that it is only enabled by default in development.

Author: Applied.

WARNING: We recommend against using this setting in production environments. It relies on Ruby's `Kernel#caller` method which tends to allocate a lot of memory in order to generate stacktraces of method calls. Use query log tags (see below) instead.

### Verbose Enqueue Logs

Similar to the "Verbose Query Logs" above, allows to print source locations of methods that enqueue background jobs.

Verbose enqueue logs are enabled by default in the development environment logs after Rails 7.1.

```rb

> # config/application.rb
> # or any environment file:

Reviewer 2: ```suggestion
# config/application.rb
```

Author: applied

logging slow queries (e.g. [MySQL](https://dev.mysql.com/doc/refman/en/slow-query-log.html), [PostgreSQL](https://www.postgresql.org/docs/current/runtime-config-logging.html#GUC-LOG-MIN-DURATION-STATEMENT)),
viewing currently running queries, or for end-to-end tracing tools.

```rb

> # config/application.rb
> # or any environment file:

Reviewer 2: ```suggestion
```

I think the original version was fine.

Author: Change applied.
